README.md
# 🚧 Nelda: Experimental fork of Selda
**WARN: WIP**

[Selda](https://github.com/valderman/selda) is a Haskell library for interacting with SQL-based relational databases.
Compared to other haskell type-safe SQL libraries, for example Beam and Opaleye, `Selda` feels more lightwieght and simple. But to keep the library ease of use, there is some tradeoffs.


This fork aims experiment with a different tradeoff while trying to keeping the simplicity of `Selda` as possible.
The major differences are:

* Seperate code to implement per-DB features
* Use Extensible Record to represent (named)sets of column. Currently using [jrec](https://github.com/juspay/jrec).
* Value-level table schema definition instead of using record data type.

# Status quo

* Sample works(`selda-try`'s `Main.hs`)
* Only SQLite supported
* Some features don't work(e.g. JSON)

# How it looks like

This is how Selda's [sample](https://selda.link/) looks like with Nelda.

```haskell
data Pet = Dog | Horse | Dragon
    deriving (Show, Read, Bounded, Enum)
    deriving SqlType via SqlTypeDeriving.TextEnum Pet

people :: _
people = table #people
    ( column #name T.text & notNull & primary
    , column #age  T.int  & notNull
    , column #pet  (T.text & asSqlType @Pet) & default_ Dog
    )

test :: IO _
test = withSQLite "people.sqlite" $ do
    createTable people

    insert_ people
        [ Rec (#name := ("Velvet" :: Text), #age := (19 :: Int), #pet := Just Dog)
        , Rec (#name := "Kobayashi",        #age := 23,          #pet := Just Dragon)
        , Rec (#name := "Miyu",             #age := 10,          #pet := Nothing)
        ]

    query $ do
        row <- select people
        restrict $ row.age .>= 18
        pure row
```

# Motivation
## Implemeting perDB features by seperating code

### NOTE: Backpack
I was first trying to use [Backpack](https://gitlab.haskell.org/ghc/ghc/-/wikis/backpack),
but stop using since it didn't support recursive component,
and lack of supports from build tools(e.g. `stack`) and `HLS`/`ghcide`.

## Utilize Extensible Record(jrec)
## Value-level table schema definition
# Other difference from Selda
## Dropped inductive tuple and supported normal tuple instead

Selda uses *Inductive Tuple* (e.g. `a :*: b`) at the boundary of `JOIN`, `AGGREGATE` and `QUERY`.
Inductive tuples can have arbitrary length unlike normal tuples where you have limit of 62.
Also, this a internal thing but, implementing type class instances or type families for inductive tuple is quite simple and clean
compared to normal tuple where you need definition for each tuple size you want to support.

But there are some drawbacks:

* Syntax noiseness.
`a :*: b :*: c` is noisy and unnatural compared to `(a, b, c)`.
* Can only take values out by pattern matching.
Pattern matching against tuples larger than certain size starts to become error-prone,
especialy when the values tend to have same types.

Therefore, Nelda dropped inductive tuple support and instead supported normal tuple(e.g. `(a, b)`) upto 8-tuple.
If you want more than 8 `Col s a`s to cross the boundary, explicity name each `Col s a` and build (or add to existing) `Row s a` is recomended way
(or you can use nested tuples, but then you have the same problem as inductive tuple).

# Things TODO

* [x] ENUM support(TEXT type backend)
* [X] Table creation
* [X] Add more column-level constraints/attirbutes
* [ ] Add Table-level constraints/attributes
* [ ] MySQL backend support
* [ ] PostgreSQL backend support
* [ ] Add more columns type
* [ ] Table validation
* [ ] JP -> EN (README, comments)
* [ ] Table migration
* [ ] JSON support
* [ ] ENUM support(PostgreSQL)
* [ ] Tests


nelda-try/CHANGELOG.md
# Revision history for selda-try

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


selda-try/Main.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingVia #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PartialTypeSignatures #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -fplugin=RecordDotPreprocessor #-}

module Main where

import Database.Nelda.SQLite (withSQLite)
import Database.Nelda.Schema as S hiding (Columns)
import Database.Nelda.Schema.ColumnType as T (int, text)

-- import Database.Nelda.Schema.ColumnConstraint
import Database.Nelda.SqlTypeDeriveStrategy as SqlTypeDeriving (
    TextEnum (..),
 )

-- import qualified Database.Nelda.Action as Nelda
-- import qualified Database.Nelda.Compile.Table as CreateTable
-- import qualified Database.Nelda.Compile.Index as CreateIndex

import Data.Function ((&))
import Data.Text (Text)
import Text.Pretty.Simple

import Database.Nelda.Action (query)
import Database.Nelda.Query.SqlClause (innerJoin, leftJoin, select)
import Database.Nelda.Query.SqlExpression
import Database.Nelda.SQL.Col (Col)
import Database.Nelda.SQL.Nullability as N
import Database.Nelda.SQL.Row (dropCol, emptyRow, (!+))
import Database.Nelda.SQL.RowHasFieldInstance ()
import Database.Nelda.SqlType (SqlType)
import GHC.Generics (Generic)
import GHC.OverloadedLabels (IsLabel (fromLabel))
import GHC.Records.Compat (HasField)
import JRec
import JRec.Internal (FldProxy (..), Has, Set, get, set)

-- * HasField(record-hasfield) instance for Rec(jrec)

instance (Has l lts t, Set l lts t ~ lts) => HasField l (Rec lts) t where
    hasField r =
        ( \t -> set (FldProxy :: FldProxy l) t r
        , get (FldProxy :: FldProxy l) r
        )

-- | Evil use of HasField/RecordDotSyntax
--
-- With this instance, all functions that recieves OverloadedLabels as first arugment
-- can accept their first argument thourgh RecordDotSyntax.
--
-- For example:
--
--   column #age intType
--
-- could be written
--
--   column.age intType
--
-- NOTE: You shouldn't use this.
-- Use only if you can't bear the syntax noiseness of OverloadedLabels.
instance (IsLabel label a, b ~ b') => HasField label ((->) a b) b' where
    hasField f =
        ( \b -> const b <$> f -- Type-checks. But does it make sense?
        , f (fromLabel @label @a)
        )

-- | One more evil case.
-- instance (HasField label a b, b' ~ Maybe b) => HasField label (Mabye a) b'

-- * Sample

data Pet = Dog | Horse | Dragon
    deriving (Show, Read, Bounded, Enum)
    deriving (SqlType) via SqlTypeDeriving.TextEnum Pet

people =
    table
        #people
        ( column #name T.text & notNull & primary
        , column #age T.int & notNull
        , column #pet (T.text & asSqlType @Pet) & default_ Dog
        )
        & addIndex (Single #name)

data People2 = People2
    { name :: Text
    , age :: Int
    }
    deriving (Generic)

-- sp :: Row _ _ _ -> Col _ 'N.Nullable Pet
-- sp p = p.pet

pred' :: Col s 'NonNull Text -> Col s 'NonNull Bool
pred' _a = undefined

test = withSQLite "people.sqlite" $ do
    -- liftIO $ print $ CreateTable.compileCreateTable CreateTable.defaultConfig people
    -- let Table{tabIndexies} = people
    -- liftIO $ print $ map (CreateIndex.compileCreateIndex CreateIndex.defaultConfig) tabIndexies
    -- createTable people

    -- insert_ people
    --     [ Rec (#name := ("Velvet" :: Text), #age := (19 :: Int), #pet := Just Dog)
    --     , Rec (#name := "Kobayashi", #age := 23, #pet := Just Dragon)
    --     , Rec (#name := "Miyu",      #age := 10, #pet := Nothing)
    --     ]
    -- insertFromNative_ people
    --     [ People2 "foo" 23 ]

    query $ do
        row <- select people
        -- val <- values ([] :: [Rec '["hoge" := Pet]])
        -- restrict $ row.age .>= 18
        -- 引数の順序変えたらエラーが出るところが変わった...
        -- 推論の何かが影響しているのか...
        _ <-
            leftJoin
                (\a -> a.name .== "hoge")
                ( do
                    p <- select people
                    -- pure $ null_ @Bool
                    pure p
                )

        i <- innerJoin (\a -> a) (pure true_)
        let a =
                emptyRow
                    & #foo !+ i
                    & #hoge !+ true_
                    & #foo2 !+ row.name
                    & dropCol #hoge
        -- pure (row.age, row.name, row.pet, i)
        pure (i, a)

-- TODO: pPrint の出力がコンパクトになるように調整したい
main :: IO ()
main = do
    -- pPrint $ people
    -- pPrint $ compQuery 0 $ do
    --     Nelda.select people
    -- pPrint $ compQuery 0 $ do
    --     row <- Nelda.select people
    --     restrict $ row ! #age .== 3
    --     pure row
    pPrint =<< test
    -- pPrint $ bar
    -- pPrint $ people
    -- pPrint $ compQuery 0 $ do
    --     row <- select people
    --     restrict $ row ! #age .== 3
    --     pure row
    -- pPrint $ compile 0 $ do
    --     selectValues [ Person "a" 4 Nothing]
    putStrLn "Hello, Haskell!"
